mlflow/langchain/__init__.py

```diff
@@ -1003,37 +1005,62 @@ def autolog(
             new model version of the registered model with this name.
             The registered model is created if it does not already exist.
         extra_tags: A dictionary of extra tags to set on each managed run created by autologging.
+        extra_log_classes: A list of langchain classes to log in addition to the default classes.
+            We do not guarantee classes specified in this list can be logged as a model, but tracing
+            will be supported. Note that all classes within the list must be subclasses of Runnable,
+            and we only patch `invoke`, `batch`, and `stream` methods for tracing.
     """
-
     with contextlib.suppress(ImportError):
+        import langchain
+        import langchain_community
         from langchain.agents.agent import AgentExecutor
         from langchain.chains.base import Chain
         from langchain.schema import BaseRetriever
-
-        classes = lc_runnables_types() + (AgentExecutor, Chain)
-        for cls in classes:
-            # If runnable also contains loader_fn and persist_dir, warn
-            # BaseRetrievalQA, BaseRetriever, ...
-            safe_patch(
-                FLAVOR_NAME,
-                cls,
-                "invoke",
-                functools.partial(patched_inference, "invoke"),
-            )
-
-            safe_patch(
-                FLAVOR_NAME,
-                cls,
-                "batch",
-                functools.partial(patched_inference, "batch"),
-            )
-
-            safe_patch(
-                FLAVOR_NAME,
-                cls,
-                "stream",
-                functools.partial(patched_inference, "stream"),
-            )
+        from langchain.schema.runnable import Runnable
+
+        # avoid duplicate patching
+        patched_class = set()
+        # avoid infinite recursion
+        inspected_module = set()
+
+        def _patch_runnable_cls(cls):
```

Comment:
can this function be moved to the top level?

Reply:
You mean outside of the contextlib.suppress(ImportError) or outside of autolog? I want it to be only accessible by autolog though to avoid any misusage on it